File: python-cli/src/convex_notes/entity/note.py
"""Note entity models."""

from datetime import datetime
from typing import Self

from pydantic import BaseModel, Field


class Note(BaseModel):
    """A note entity representing a stored note."""

    id: str = Field(description="Unique identifier")
    title: str = Field(description="Note title")
    content: str = Field(description="Note content")
    created_at: int = Field(description="Creation timestamp (Unix milliseconds)")
    updated_at: int = Field(description="Last update timestamp (Unix milliseconds)")

    @property
    def short_id(self) -> str:
        """Get a short ID for display (first 12 chars)."""
        return self.id[:12] if len(self.id) > 12 else self.id

    @property
    def content_preview(self) -> str:
        """Get a content preview (max 40 chars)."""
        if len(self.content) > 40:
            return f"{self.content[:40]}..."
        return self.content

    @property
    def created_datetime(self) -> datetime:
        """Get creation time as datetime."""
        return datetime.fromtimestamp(self.created_at / 1000)

    @property
    def formatted_created(self) -> str:
        """Get formatted creation date string."""
        return self.created_datetime.strftime("%Y-%m-%d %H:%M")

    @classmethod
    def from_convex(cls, data: dict) -> Self:
        """Create a Note from Convex response data."""
        return cls(
            id=data.get("_id", ""),
            title=data.get("title", ""),
            content=data.get("content", ""),
            created_at=int(data.get("createdAt", 0)),
            updated_at=int(data.get("updatedAt", 0)),
        )


class CreateNote(BaseModel):
    """Data for creating a new note."""

    title: str = Field(min_length=1, description="Note title")
    content: str = Field(min_length=1, description="Note content")


class UpdateNote(BaseModel):
    """Data for updating an existing note."""

    id: str = Field(description="Note ID to update")
    title: str = Field(min_length=1, description="New title")
    content: str = Field(min_length=1, description="New content")

File: backend/crates/db/src/entity/note.rs
use serde::{Deserialize, Serialize};

/// A note entity representing a stored note.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Note {
    /// Unique identifier
    pub id: String,
    /// Note title
    pub title: String,
    /// Note content
    pub content: String,
    /// Creation timestamp (Unix milliseconds)
    pub created_at: i64,
    /// Last update timestamp (Unix milliseconds)
    pub updated_at: i64,
}

impl Note {
    /// Get a short ID for display (first 12 chars)
    pub fn short_id(&self) -> &str {
        if self.id.len() > 12 {
            &self.id[..12]
        } else {
            &self.id
        }
    }

    /// Get a content preview (max 40 chars)
    pub fn content_preview(&self) -> String {
        if self.content.len() > 40 {
            format!("{}...", &self.content[..40])
        } else {
            self.content.clone()
        }
    }
}

/// Data for creating a new note.
#[derive(Debug, Clone)]
pub struct CreateNote {
    pub title: String,
    pub content: String,
}

impl CreateNote {
    pub fn new(title: impl Into<String>, content: impl Into<String>) -> Self {
        Self {
            title: title.into(),
            content: content.into(),
        }
    }

    /// Validate the note data.
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.title.trim().is_empty() {
            return Err("Title is required");
        }
        if self.content.trim().is_empty() {
            return Err("Content is required");
        }
        Ok(())
    }
}

/// Data for updating an existing note.
#[derive(Debug, Clone)]
pub struct UpdateNote {
    pub id: String,
    pub title: String,
    pub content: String,
}

impl UpdateNote {
    pub fn new(
        id: impl Into<String>,
        title: impl Into<String>,
        content: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            title: title.into(),
            content: content.into(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_note_short_id() {
        let note = Note {
            id: "abc123def456xyz789".to_string(),
            title: "Test".to_string(),
            content: "Content".to_string(),
            created_at: 0,
            updated_at: 0,
        };
        assert_eq!(note.short_id(), "abc123def456");
    }

    #[test]
    fn test_content_preview() {
        let note = Note {
            id: "1".to_string(),
            title: "Test".to_string(),
            content: "A very long content that exceeds forty characters limit".to_string(),
            created_at: 0,
            updated_at: 0,
        };
        assert!(note.content_preview().ends_with("..."));
        assert!(note.content_preview().len() <= 43);
    }

    #[test]
    fn test_create_note_validation() {
        let valid = CreateNote::new("Title", "Content");
        assert!(valid.validate().is_ok());

        let empty_title = CreateNote::new("", "Content");
        assert!(empty_title.validate().is_err());

        let empty_content = CreateNote::new("Title", "   ");
        assert!(empty_content.validate().is_err());
    }
}
